src/meta-srv/src/procedure/create_table.rs

```diff
@@ -0,0 +1,329 @@
+
+        let table_route_key = TableRouteKey::with_table_name(table_id, &table_name.clone().into())
+            .key()
+            .into_bytes();
+
+        let table_global_key = TableGlobalKey {
+            catalog_name: table_name.catalog_name.clone(),
+            schema_name: table_name.schema_name.clone(),
+            table_name: table_name.table_name.clone(),
+        }
+        .to_string()
+        .into_bytes();
+
+        let (peers, table_route) = self
+            .creator
+            .data
+            .table_route
+            .clone()
+            .try_into_raw()
+            .context(error::ConvertProtoDataSnafu)?;
+
+        let table_route_value = TableRouteValue {
+            peers,
+            table_route: Some(table_route),
+        };
+
+        let table_global_value = create_table_global_value(
+            &table_route_value,
+            self.creator.data.task.table_info.clone(),
+        )?
+        .as_bytes()
+        .context(error::InvalidCatalogValueSnafu)?;
+
+        let txn = Txn::new()
+            .when(vec![
+                Compare::with_not_exist_value(table_route_key.clone(), CompareOp::Equal),
+                Compare::with_not_exist_value(table_global_key.clone(), CompareOp::Equal),
+            ])
+            .and_then(vec![
+                TxnOp::Put(table_route_key, table_route_value.into()),
+                TxnOp::Put(table_global_key, table_global_value),
+            ]);
+
+        let resp = self.context.kv_store.txn(txn).await?;
+
+        ensure!(
+            resp.succeeded,
+            error::TxnSnafu {
+                msg: "table_route_key or table_global_key exists"
+            }
+        );
+
+        Ok(())
+    }
+
+    async fn on_create_metadata(&mut self) -> Result<Status> {
+        let kv_store = &self.context.kv_store;
+        let key = &self.global_table_key();
+
+        match get_table_global_value(kv_store, key).await? {
+            Some(table_global_value) => {
+                // The metasrv crashed after metadata was created immediately.
+                // Recovers table_route from kv.
+                let table_id = table_global_value.table_id() as u64;
+
+                let expected = self.creator.data.table_route.table.id;
+                // If there is something like:
+                // Create table A, Create table A(from another Fe, Somehow, Failed), Renames table A to B, Create table A(Recovered).
+                // We must ensure the table_id isn't changed.
+                ensure!(
+                    table_id == expected,
+                    error::TableIdChangedSnafu {
+                        expected,
+                        found: table_id
+                    }
+                );
+            }
+            None => {
+                // registers metadata
+                self.register_metadata().await?;
+            }
+        }
+
+        self.creator.data.state = CreateTableState::DatanodeCreateTable;
```

Comment:
What happens if a user makes a request when the metadata and routing information are complete, but the table creation is not yet finished?

Reply:
I moved the `CreateMetadata` step after the `DatanodeCreateTable`.